Add AMD GPU collector

Dim-P opened the pull request:
##### Summary

This PR adds a collector that monitors GPU and VRAM utilization, frequency and usage of all AMD GPU cards on the system.

Fixes https://github.com/netdata/netdata/issues/12616.

##### Test Plan

TODO

##### Additional Information
<!-- This is usually used to help others understand your
motivation behind this change. A step-by-step reproduction of the problem is
helpful if there is no related issue. -->

<details> <summary>For users: How does this change affect me?</summary>
  <!--
Describe the PR affects users: 
- Which area of Netdata is affected by the change?
- Can they see the change or is it an under the hood? If they can see it, where?
- How is the user impacted by the change? 
- What are there any benefits of the change? 
-->
</details>

ilyam8 commented on collectors/proc.plugin/multi_metadata.yaml:
```diff
@@ -5180,3 +5180,105 @@ modules:
+          labels:
+            - name: product_name
+              description: GPU product name (e.g. AMD RX 6600)
+          metrics:
+            - name: amdgpu.utilization
+              description: GPU utilization
+              unit: "percentage"
+              chart_type: line
+              dimensions:
+                - name: GPU
+                - name: memory
```
These should be different charts. The same goes for other charts. Think of a chart as a metric:

 - gpu_utilization
 - gpu_memory_utilization

In general, if dimensions are not summable - you need to put them in different charts.

Dim-P replied:
I just saw that's how it is implemented in `nvidia-smi`. Is that so they are aggregatable per label?

ilyam8 replied:
nvidia smi doesn't mix different metrics in one chart, or what do you mean?

Dim-P replied:
Yeah, I meant it is implemented as you said but I didn't see that before your comment.

Dim-P commented:
@ilyam8 Removed configuration as discussed. Please let me know if chart IDs and contexts make sense as they are, I think they do now. I tested with multiple GPU cards and aggregated metrics seem to work correctly.

ilyam8 commented on collectors/proc.plugin/sys_class_drm.c:
```diff
@@ -0,0 +1,1076 @@
+
+    }
+
+    return 0;
```
I think we should return 1 if no cards are discovered.

Dim-P replied:
Good point, I will change the code. I also need to rebase since `multi_metadata.yaml` is no longer a thing.

Dim-P commented:
> Hi Dimitri! Nice!
> 
> So, I tested this on my laptop, which has an AMD 4500U CPU, with integrated GPU.
> 
> The card is on this [list](https://www.coelacanth-dream.com/posts/2019/12/30/did-rid-product-matome-p2/#renoir-gfx90c).
> 
> I've added `{0x1636, 0xC3, "AMD 4500U"},` in your list, since it wasn't included. So maybe there are some more cards that could be added, but that can be iterated.
> 
> What I've found next, is that still it wouldn't display any charts for it. The problem was that in my case `util_mem` does not exist. Commenting out the part for checking `util_mem`, did allow other charts to appear (gpu_utilization, gpu_clk_frequency and gpu_mem_clk_frequency).
> 
> So a suggestion would be that if in case something is missing, don't give up on the whole card?
> 
> Also, if possible to capitalize the `amdgpu` entry in the list on the dashboard (e.g. `AMDGpu`) and an icon would be cool as well :-)

Thanks, good points. 

About additional GPU marketing names, I believe these can be added later, as it's a never-ending piece of work. I did add some RX 7xxx though that were missing from the original list.

I changed how chart updates are performed and now a linked list is used, so any metrics (and charts) that cannot be monitored any more can easily be removed from the list. This seems to be working fine but please test on your system again.

About `amdgpu`, it is capitalized if you check the first version of the dashboard (`localhost:19999/v1`), but I believe for `v2`, this (and the icon) is handled by @netdata/cloud-fe .

I also checked @Ferroin 's concerns about introducing unnecessary GPU usage due to `/sys/class/drm` calls, but I didn't notice any difference on my setup.

MrZammler reviewed (approved):
Seems to work fine here! Just needs more cards, we can add later!

ilyam8 commented on collectors/proc.plugin/sys_class_drm.c:
```diff
@@ -0,0 +1,1148 @@
+
+                snprintfz(filename, FILENAME_MAX, "%s/%s", c->pathname, "device/device");
+                if(read_single_base64_or_hex_number_file(filename, (unsigned long long *) &c->id.asic_id)){
+                    collector_error("Cannot read asic_id from '%s'", filename);
```
I think all these errors are not errors. But debug or info at least.

Having AMD GPU is absolutely not required so failing to read those files is not an error but expected.

If we keep them as it is **all** users w/o AMD GPU will get potentially NumOfProps*NumberOfCards (in `/sys/class/drm`) error log lines on every restart.

ilyam8 commented on collectors/proc.plugin/sys_class_drm.c:
```diff
@@ -0,0 +1,1148 @@
+            if(de->d_type == DT_LNK && !strncmp(de->d_name, "card", 4) && !strchr(de->d_name, '-')) {
+                struct card *const c = callocz(1, sizeof(struct card));  
+
+                /* Get static info */              
```
All AMD cards have `DRIVER=amdgpu` in `/sys/class/drm/cardX/device/uevent`. I think you need to check this, it will remove log messages like on every (re)start

```
2023-08-07 14:42:05: netdata INFO  : P[proc] : Cannot read file '/sys/class/drm/card0/device/gpu_busy_percent'
2023-08-07 14:42:05: netdata INFO  : P[proc] : Cannot read file '/sys/class/drm/card0/device/mem_busy_percent'
2023-08-07 14:42:05: netdata INFO  : P[proc] : Cannot read file '/sys/class/drm/card0/device/pp_dpm_sclk'
2023-08-07 14:42:05: netdata INFO  : P[proc] : Cannot read file '/sys/class/drm/card0/device/pp_dpm_mclk'
2023-08-07 14:42:05: netdata INFO  : P[proc] : Cannot read file '/sys/class/drm/card0/device/mem_info_vram_used'
2023-08-07 14:42:05: netdata INFO  : P[proc] : Cannot read file '/sys/class/drm/card0/device/mem_info_vram_total'
2023-08-07 14:42:05: netdata INFO  : P[proc] : Cannot read file '/sys/class/drm/card0/device/mem_info_vis_vram_used'
2023-08-07 14:42:05: netdata INFO  : P[proc] : Cannot read file '/sys/class/drm/card0/device/mem_info_vis_vram_total'
2023-08-07 14:42:05: netdata INFO  : P[proc] : Cannot read file '/sys/class/drm/card0/device/mem_info_gtt_used'
```

Dim-P replied:
Done

thiagoftsm commented on collectors/proc.plugin/metadata.yaml:
```diff
@@ -5180,3 +5180,138 @@ modules:
+    overview:
+      data_collection:
+        metrics_description: "This integration monitors AMD GPU metrics, such as utilization, clock frequency and memory usage."
+        method_description: ""
```
Please, add a method description.

Dim-P replied:
What's a method description? None of the other `proc` collectors has one.

ilyam8 replied:
Let's update meta in a separate PR.

thiagoftsm commented on collectors/proc.plugin/metadata.yaml:
```diff
@@ -5180,3 +5180,138 @@ modules:
+      availability: []
+      scopes:
+        - name: gpu
+          description: "These metrics refer to the GPU."
```
I suggest to also add `AMD` word to description considering PR's name, or is it possible to use the code with NVIDIA also?

Dim-P replied:
I literally copied this from https://github.com/netdata/go.d.plugin/blob/430340af44381a1651b0175064b781510c96f262/modules/nvidia_smi/metadata.yaml#L115 .

thiagoftsm commented on libnetdata/inlined.h:
```diff
@@ -510,6 +510,21 @@ static inline int read_single_signed_number_file(const char *filename, long long
 }
 
+static inline int read_single_base64_or_hex_number_file(const char *filename, unsigned long long *result) {
+    char buffer[30 + 1];
```
Please, can you create a MACRO or to use an existing instead to use number `30`?

Dim-P replied:
I kept it the same as in the other `read_single_X` functions`.

thiagoftsm commented on web/gui/dashboard_info.js:
```diff
@@ -140,6 +140,12 @@ netdataDashboard.menu = {
+        title: 'AMD GPUs',
+        icon: '<i class="fas fa-microchip"></i>',
+        info: 'Detailed information for each AMD GPU of the system. Temperature, fan speed, voltage and power metrics can be found at the <a href="#menu_sensors">Sensors</a> section.'
+    },
```
Please, do not forget to create an analogous PR to also bring this description for new dashboard.

ilyam8 replied:
Minor: we shouldn't update this file in netdata/netdata but in https://github.com/netdata/dashboard. If there is a new version of the old dashboard, this file will be overwritten.

thiagoftsm reviewed (approved):
I do not have hardware to test, bu considering that other teammate already tested it, and possible changes will come in other PRs, LGTM!
